akka-actor/src/main/scala/akka/io/dns/internal/TcpDnsClient.scala

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (C) 2018 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.io.dns.internal
+
+import java.net.InetSocketAddress
+
+import akka.AkkaException
+import akka.actor.{ Actor, ActorLogging, ActorRef, Stash }
+import akka.annotation.InternalApi
+import akka.io.Tcp._
+import akka.io.dns.internal.DnsClient.{ Answer, DnsQuestion, Question4 }
+import akka.io.{ IO, Tcp }
+import akka.util.ByteString
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] class TcpDnsClient(ns: InetSocketAddress, answerRecipient: ActorRef) extends Actor with ActorLogging with Stash {
+  import TcpDnsClient._
+
+  import context.system
+
+  override def receive: Receive = idle
+
+  val tcp = IO(Tcp)
+
+  val idle: Receive = {
+    case _: Message ⇒
+      stash()
+      log.debug("Connecting to [{}]", ns)
+      tcp ! Tcp.Connect(ns)
+      context.become(connecting)
+  }
+
+  val connecting: Receive = {
+    case CommandFailed(_: Connect) ⇒
+      log.warning("Failed to connect to [{}]", ns)
+      throw new AkkaException("Connecting failed")
+    case _: Tcp.Connected ⇒
+      log.debug(s"Connected to TCP address [{}]", ns)
+      val connection = sender()
+      context.become(ready(connection))
+      connection ! Register(self)
+      unstashAll()
+    case _: Message ⇒
+      stash()
+  }
+
+  def ready(connection: ActorRef, buffer: ByteString = ByteString.empty): Receive = {
+    case msg: Message ⇒
+      val bytes = msg.write()
+      connection ! Tcp.Write(encodeLength(bytes.length))
+      connection ! Tcp.Write(bytes)
+    case CommandFailed(_: Write) ⇒
+      log.warning("Write failed")
+      throw new AkkaException("Write failed")
+    case Received(newData) ⇒
+      val data = buffer ++ newData
+      // TCP DNS responses are prefixed by 2 bytes encoding the length of the response
+      val prefixSize = 2
+      if (data.length < prefixSize)
+        context.become(ready(connection, data))
+      else {
+        val expectedPayloadLength = decodeLength(data)
+        if (data.drop(prefixSize).length < expectedPayloadLength)
+          context.become(ready(connection, data))
+        else {
+          val payload = data.drop(prefixSize).take(expectedPayloadLength)
```

Comment:
Isn't there something with drop and take on ByteStrings being overly expensive?

Reply:
Since we're doing I/O anyway and we need to seek in the message while parsing references to domain names, I think the `drop` is worth it. The `take` we can skip, though.